### PR TITLE
Mob Abilities

### PR DESCRIPTION
**Pull request by Whoneedspacee**
## About The Pull Request

[Video Showcase](https://www.youtube.com/watch?v=fE9AgxZXbPs)

This overhauls the cooldown action subtype to allow for shared cooldowns and for targeted abilities where you can queue up an ability and then click a target.

It's not a perfect system but it's way better than what is there currently by a longshot, and allows you to for example, add a charging action to any mob, not just simplemobs.

## Why It's Good For The Game

Allows for more easy mobs to be made or even allows for some custom mob creation by admins mid-round as opposed to most of these abilities just being hard-coded procs on the mobs.

## Changelog

:cl: Whoneedspacee
refactor: Refactors megafauna abilities to be mob actions which are grantable to any living mob.
admin: Added a verb under admin events which lets you grant mob actions to a marked living mob.
/:cl:

### CODE REVIEW

File: code/__DEFINES/dcs/signals/signals_mob/signals_mob_simple.dm [hunk 1]
```diff
@@ -13,3 +13,18 @@
...
+	#define COMPONENT_OVERRIDE_CHARGE_BUMP (1<<0)
+#define COMSIG_FINISHED_CHARGE "mob_ability_charge_finished"
+#define COMSIG_PROJECTILE_FIRING_STARTED "mob_ability_started_projectile"
+#define COMSIG_PROJECTILE_FIRING_FINISHED "mob_ability_fired_projectile"
+#define COMSIG_SPIRAL_ATTACK_START "mob_spiral_attack_start"
+#define COMSIG_SPIRAL_ATTACK_FINISHED "mob_spiral_attack_finished"
+#define COMSIG_SWOOP_ATTACK_STARTED "mob_swoop_attack_started"
+#define COMSIG_SWOOP_INVULNERABILITY_STARTED "mob_swoop_invulnerability_started"
+#define COMSIG_SWOOP_ATTACK_FINISHED "mob_swoop_attack_finished"
+#define COMSIG_LAVA_ARENA_FAILED "mob_lava_arena_failed"
```
Comment (Watermelon914) [inline code/__DEFINES/dcs/signals/signals_mob/signals_mob_simple.dm:30]: needs documenting

File: code/datums/action.dm [hunk 2]
```diff
@@ -673,23 +681,70 @@
...
 	UpdateButtonIcon()
 	START_PROCESSING(SSfastprocess, src)
 
-/datum/action/cooldown/process()
-	if(!owner)
-		button.maptext = ""
-		STOP_PROCESSING(SSfastprocess, src)
-	var/timeleft = max(next_use_time - world.time, 0)
-	if(timeleft == 0)
+/datum/action/cooldown/Trigger(atom/A)
```
Comment (Watermelon914) [inline code/datums/action.dm:700]: use more descriptive variable names

File: code/datums/actions/mobs/charge.dm [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+		return
+	if(!isliving(A))
+		return
+	var/mob/living/L = A
+	L.visible_message("<span class='danger'>[source] slams into [L]!</span>", "<span class='userdanger'>[source] tramples you into the ground!</span>")
+	source.forceMove(get_turf(L))
+	L.apply_damage(damage_dealt, BRUTE, wound_bonus = CANT_WOUND)
+	playsound(get_turf(L), 'sound/effects/meteorimpact.ogg', 100, TRUE)
+	shake_camera(L, 4, 3)
+	shake_camera(source, 2, 3)
```
Comment (Watermelon914) [inline code/datums/actions/mobs/charge.dm:146]: use more descriptive variable names for atom and living here

File: code/datums/actions/mobs/charge.dm [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+			owner.forceMove(place)
+			self_placed = TRUE
+			continue
+		var/mob/living/simple_animal/hostile/megafauna/bubblegum/hallucination/B = new /mob/living/simple_animal/hostile/megafauna/bubblegum/hallucination(place)
+		B.appearance = owner.appearance
+		B.name = "[owner]'s hallucination"
+		B.alpha = 127.5
+		B.move_through_mob = owner
+		B.spawn_blood = spawn_blood
+		INVOKE_ASYNC(src, .proc/do_charge, B, target_atom, delay, past)
```
Comment (Watermelon914) [inline code/datums/actions/mobs/charge.dm:207]: needs more descriptive of a variable name, avoid one letter variable names like these

File: code/datums/actions/mobs/dash.dm [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	dash_to(target_atom)
+	StartCooldown()
+
+/datum/action/cooldown/mob_cooldown/dash/proc/dash_to(atom/dash_target)
+	var/list/accessable_turfs = list()
+	var/self_dist_to_target = 0
+	var/turf/own_turf = get_turf(owner)
+	if(!QDELETED(dash_target))
+		self_dist_to_target += get_dist(dash_target, own_turf)
+	for(var/turf/open/O in RANGE_TURFS(dash_range, own_turf))
```
Comment (Watermelon914) [inline code/datums/actions/mobs/dash.dm:25]: rename O to something more descriptive

**Review by Watermelon914 — CHANGES_REQUESTED** [review]
A lot of variable names are 1 letter, they should ideally have a more descriptive name if you're going to refactor action code.

Since you've already got a lot of 1 letter variable names everywhere, I don't think it'd be fair to ask you to change it all, but please change the ones I've listed below

**Comment by tralezab** [comment]
i really like this decoupling

File: code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm [hunk 6]
```diff
@@ -298,14 +179,15 @@ Difficulty: Medium
...
 /mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/doom
 	name = "hostile-environment miner"
 	desc = "A miner destined to hop across dimensions for all eternity, hunting anomalous creatures."
 	speed = 8
 	move_to_delay = 8
 	ranged_cooldown_time = 0.8 SECONDS
-	dash_cooldown = 0.8 SECONDS
 
-#undef MINER_DASH_RANGE
+/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/doom/Initialize()
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm:191]:
```suggestion
/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/doom/Initialize(mapload)
```

File: code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm [hunk 7]
```diff
@@ -51,74 +47,41 @@ Difficulty: Medium
...
-
-/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/Initialize(mapload)
+	/// Dash ability
+	var/datum/action/cooldown/mob_cooldown/dash/dash
+	/// Kinetic accelerator ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/kinetic_accelerator/kinetic_accelerator
+	/// Transform weapon ability
+	var/datum/action/cooldown/mob_cooldown/transform_weapon/transform_weapon
+
+/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/Initialize()
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm:62]:
```suggestion
/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/Initialize(mapload)
```

File: code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm [hunk 8]
```diff
@@ -51,80 +53,73 @@
...
+	/// Spiral shots ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/spiral_shots/spiral_shots
+	/// Random shots ablity
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/random_aoe/random_shots
+	/// Shotgun blast ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/shotgun_blast/shotgun_blast
+	/// Directional shots ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/dir_shots/alternating/dir_shots
+
+/mob/living/simple_animal/hostile/megafauna/colossus/Initialize()
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm:66]:
```suggestion
/mob/living/simple_animal/hostile/megafauna/colossus/Initialize(mapload)
```

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm [hunk 9]
```diff
@@ -28,18 +28,49 @@
...
 	gold_core_spawnable = HOSTILE_SPAWN
+	/// Charging ability
+	var/datum/action/cooldown/mob_cooldown/charge/charge
 
-/mob/living/simple_animal/hostile/asteroid/lobstrosity/ranged_secondary_attack(atom/target, modifiers)
-	if(COOLDOWN_FINISHED(src, charge_cooldown))
-		INVOKE_ASYNC(src, /mob/living/simple_animal/hostile/.proc/enter_charge, target)
-	else
-		to_chat(src, span_notice("Your charge is still on cooldown!"))
+/mob/living/simple_animal/hostile/asteroid/lobstrosity/Initialize()
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:38]:
```suggestion
/mob/living/simple_animal/hostile/asteroid/lobstrosity/Initialize(mapload)
```

File: code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm [hunk 10]
```diff
@@ -51,74 +47,41 @@ Difficulty: Medium
...
-							   /datum/action/innate/megafauna_attack/transform_weapon)
 	move_force = MOVE_FORCE_NORMAL //Miner beeing able to just move structures like bolted doors and glass looks kinda strange
-
-/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/Initialize(mapload)
+	/// Dash ability
+	var/datum/action/cooldown/mob_cooldown/dash/dash
+	/// Kinetic accelerator ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/kinetic_accelerator/kinetic_accelerator
+	/// Transform weapon ability
+	var/datum/action/cooldown/mob_cooldown/transform_weapon/transform_weapon
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm:60]: These need to be handled in destroy

File: code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm [hunk 11]
```diff
@@ -51,80 +53,73 @@
...
-
-/mob/living/simple_animal/hostile/megafauna/colossus/Initialize(mapload)
+	/// Spiral shots ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/spiral_shots/spiral_shots
+	/// Random shots ablity
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/random_aoe/random_shots
+	/// Shotgun blast ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/shotgun_blast/shotgun_blast
+	/// Directional shots ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/dir_shots/alternating/dir_shots
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm:64]: Same here

File: code/modules/mob/living/simple_animal/hostile/megafauna/demonic_frost_miner.dm [hunk 12]
```diff
@@ -42,205 +42,76 @@ Difficulty: Extremely Hard
...
 	/// If the demonic frost miner is in its enraged state
 	var/enraged = FALSE
 	/// If the demonic frost miner is currently transforming to its enraged state
 	var/enraging = FALSE
+	/// Frost orbs ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/rapid_fire/shrapnel/frost_orbs
+	/// Snowball Machine gun Ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/rapid_fire/snowball_machine_gun
+	/// Ice Shotgun Ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/shotgun_blast/pattern/ice_shotgun
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/megafauna/demonic_frost_miner.dm:54]: same here

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm [hunk 13]
```diff
@@ -191,17 +191,48 @@
...
 	mob_size = MOB_SIZE_LARGE
 	gold_core_spawnable = NO_SPAWN
-	charger = TRUE
-	charge_distance = 4
 	menu_description = "Tank spider variant with an enormous amount of health and damage, but is very slow when not on webbing. It also has a charge ability to close distance with a target after a small windup. Does not inject toxin."
-	///Whether or not the tarantula is currently walking on webbing.
+	/// Whether or not the tarantula is currently walking on webbing.
 	var/silk_walking = TRUE
+	/// Charging ability
+	var/datum/action/cooldown/mob_cooldown/charge/charge
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm:198]: same here

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm [hunk 14]
```diff
@@ -28,18 +28,49 @@
...
 	vision_range = 5
 	aggro_vision_range = 7
-	charger = TRUE
-	charge_distance = 4
 	butcher_results = list(/obj/item/food/meat/crab = 2, /obj/item/stack/sheet/bone = 2)
 	robust_searching = TRUE
 	footstep_type = FOOTSTEP_MOB_CLAW
 	gold_core_spawnable = HOSTILE_SPAWN
+	/// Charging ability
+	var/datum/action/cooldown/mob_cooldown/charge/charge
```
Comment (MrMelbert) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:36]: same here

File: code/datums/actions/mobs/charge.dm [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+		next_move_allowed.Remove(charger)
+	UnregisterSignal(charger, COMSIG_MOVABLE_BUMP)
+	UnregisterSignal(charger, COMSIG_MOVABLE_PRE_MOVE)
+	UnregisterSignal(charger, COMSIG_MOVABLE_MOVED)
+	charging.Remove(charger)
+	SEND_SIGNAL(owner, COMSIG_FINISHED_CHARGE)
+	return TRUE
+
+/datum/action/cooldown/mob_cooldown/charge/proc/on_move(atom/source, atom/new_loc)
+	var/expected_dir = next_move_allowed[source]
```
Comment (MrMelbert) [inline code/datums/actions/mobs/charge.dm:88]:
```suggestion
/datum/action/cooldown/mob_cooldown/charge/proc/on_move(atom/source, atom/new_loc)
	SIGNAL_HANDLER

	var/expected_dir = next_move_allowed[source]
```
Also NONE of your signal procs have signal handlers as far as I can see, you should check them all

**Comment by Whoneedspacee** [comment]
update https://www.youtube.com/watch?v=ceKz0M9C8bE

File: code/__DEFINES/cooldowns.dm [hunk 16]
```diff
@@ -95,3 +95,8 @@
 #define COOLDOWN_RESET(cd_source, cd_index) cd_source.cd_index = 0
 
 #define COOLDOWN_TIMELEFT(cd_source, cd_index) (max(0, cd_source.cd_index - world.time))
+
+//Shared cooldowns for actions
```
Comment (Mothblocks) [inline code/__DEFINES/cooldowns.dm:99]: Can you put this with the rest of the cooldown keys at the top?

File: code/__DEFINES/dcs/signals/signals_mob/signals_mob_simple.dm [hunk 17]
```diff
@@ -13,3 +13,30 @@
 ///FROM mob/living/simple_animal/hostile/ooze/eat_atom(): (atom/target, edible_flags)
 #define COMSIG_OOZE_EAT_ATOM "ooze_eat_atom"
 	#define COMPONENT_ATOM_EATEN  (1<<0)
+
+//Mob ability signals /datum/action/cooldown/mob_cooldown
```
Comment (Mothblocks) [inline code/__DEFINES/dcs/signals/signals_mob/signals_mob_simple.dm:17]: Can you move a lot of these to their own file? Already not too happy about things like oozes being in here, now that I see it.

Something like signals_mob_abilities.dm or something.

Comment (Mothblocks) [reply code/__DEFINES/dcs/signals/signals_mob/signals_mob_simple.dm:17]: Also all of these need to document the proc they're called from, and what arguments are passed.

File: code/datums/action.dm [hunk 18]
```diff
@@ -673,23 +681,71 @@
 /datum/action/cooldown/IsAvailable()
 	return next_use_time <= world.time
 
-/datum/action/cooldown/proc/StartCooldown()
-	next_use_time = world.time + cooldown_time
-	button.maptext = MAPTEXT("<b>[round(cooldown_time/10, 0.1)]</b>")
+/// Starts a cooldown time, will use default cooldown time if an override is not specified, shares that cooldown with similar abilities
+/datum/action/cooldown/proc/StartCooldown(override_cooldown_time, will_share_cooldowns = TRUE)
+	if(shared_cooldown && will_share_cooldowns)
+		for(var/datum/action/cooldown/C in owner.actions - src)
```
Comment (Mothblocks) [inline code/datums/action.dm:687]: Better name

File: code/datums/action.dm [hunk 19]
```diff
@@ -673,23 +681,71 @@
 /datum/action/cooldown/IsAvailable()
 	return next_use_time <= world.time
 
-/datum/action/cooldown/proc/StartCooldown()
-	next_use_time = world.time + cooldown_time
-	button.maptext = MAPTEXT("<b>[round(cooldown_time/10, 0.1)]</b>")
+/// Starts a cooldown time, will use default cooldown time if an override is not specified, shares that cooldown with similar abilities
+/datum/action/cooldown/proc/StartCooldown(override_cooldown_time, will_share_cooldowns = TRUE)
```
Comment (Mothblocks) [inline code/datums/action.dm:685]: The boolean parameter shouldn't be here, this should just be two separate procs, with one being the process after this that starts the cooldown, and nothing else.

The two places in which you actually use `will_share_cooldowns = FALSE` you would call that directly. `StartCooldown()` itself would run the shared_cooldown stuff, though.

File: code/datums/action.dm [hunk 20]
```diff
@@ -673,23 +681,71 @@
...
+		if(target)
+			// For automatic / mob handling
+			return InterceptClickOn(owner, null, target)
+		if(owner.click_intercept == src)
+			owner.click_intercept = null
+			to_chat(owner, "<b>[src] is no longer active</b>")
+		else
+			owner.click_intercept = src
+			to_chat(owner, "<b>You are now prepared to use [src]</b>")
+		for(var/datum/action/cooldown/C in owner.actions)
```
Comment (Mothblocks) [inline code/datums/action.dm:716]: Better name

File: code/datums/action.dm [hunk 21]
```diff
@@ -673,23 +681,71 @@
...
+		return
 	if(!owner)
+		return FALSE
+	if(click_to_activate)
+		if(target)
+			// For automatic / mob handling
+			return InterceptClickOn(owner, null, target)
+		if(owner.click_intercept == src)
+			owner.click_intercept = null
+			to_chat(owner, "<b>[src] is no longer active</b>")
```
Comment (Mothblocks) [inline code/datums/action.dm:712]:
```suggestion
			to_chat(owner, "<b>[src] is no longer active.</b>")
```

In an ideal world these are not necessary at all, by the way. The icon itself should have some indication to whether or not its active. If it does, then these do not need to exist.

File: code/datums/action.dm [hunk 22]
```diff
@@ -673,23 +681,71 @@
...
+		else
+			owner.click_intercept = src
+			to_chat(owner, "<b>You are now prepared to use [src]</b>")
```
Comment (Mothblocks) [inline code/datums/action.dm:715]:
```suggestion
			to_chat(owner, "<b>You are now prepared to use [src].</b>")
```

File: code/datums/action.dm [hunk 23]
```diff
@@ -673,23 +681,71 @@
...
+		for(var/datum/action/cooldown/C in owner.actions)
+			C.UpdateButtonIcon()
+		return TRUE
+	return Activate(owner)
+
+/datum/action/cooldown/proc/InterceptClickOn(mob/living/caller, params, atom/A)
```
Comment (Mothblocks) [inline code/datums/action.dm:721]: Better variable name for `A` (such as `target`)

File: code/datums/action.dm [hunk 24]
```diff
@@ -673,23 +681,71 @@
...
+	if(!IsAvailable())
+		return FALSE
+	if(!A)
+		return FALSE
+	Activate(A)
+	caller.click_intercept = null
+	return TRUE
+
+/datum/action/cooldown/proc/Activate(atom/target_atom)
```
Comment (Mothblocks) [inline code/datums/action.dm:730]: Document

Comment (Mothblocks) [reply code/datums/action.dm:730]: Also this can just be `target`, `atom` is already in its typepath anyway.

File: code/datums/action.dm [hunk 25]
```diff
@@ -673,23 +681,71 @@
...
+	return
+
+/datum/action/cooldown/UpdateButtonIcon(status_only = FALSE, force = FALSE)
+	. = ..()
+	var/time_left = max(next_use_time - world.time, 0)
+	if(button)
+		if(text_cooldown)
+			button.maptext = "<b>[round(time_left/10, 0.1)]</b>"
```
Comment (Mothblocks) [inline code/datums/action.dm:738]:
```suggestion
			button.maptext = MAPTEXT("<b>[round(time_left/10, 0.1)]</b>")
```

Maptext without this looks fairly bad.

File: code/datums/action.dm [hunk 26]
```diff
@@ -673,23 +681,71 @@
...
+	if(!owner || time_left == 0)
 		button.maptext = ""
+	if(IsAvailable() && owner.click_intercept == src)
+		button.color = rgb(0,255,0)
```
Comment (Mothblocks) [inline code/datums/action.dm:742]:
```suggestion
		button.color = COLOR_GREEN
```

?

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Going to just assume the other stuff isn't yours, though there's plenty upsetting about them

File: code/datums/action.dm [hunk 27]
```diff
@@ -671,25 +679,84 @@
 	button.maptext_height = 12
 
 /datum/action/cooldown/IsAvailable()
-	return next_use_time <= world.time
-
-/datum/action/cooldown/proc/StartCooldown()
-	next_use_time = world.time + cooldown_time
-	button.maptext = MAPTEXT("<b>[round(cooldown_time/10, 0.1)]</b>")
+	return (..() && (next_use_time <= world.time))
```
Comment (Mothblocks) [inline code/datums/action.dm:682]:
```suggestion
	return ..() && (next_use_time <= world.time)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Just one thing but I'll self commit it

**Pull request merged**